src/Microsoft.Azure.ServiceBus.KeyVault/KeyVaultPlugin.cs

```diff
@@ -0,0 +1,195 @@
+            {
+                if (message.UserProperties.ContainsKey(Constants.InitializationVectorPropertyName) || message.UserProperties.ContainsKey(Constants.KeyNamePropertyName))
+                {
+                    return message;
+                }
+
+                var iV = await KeyVaultPlugin.GenerateInitializationVector();
+                var secret = await secretManager.GetHashedSecret(secretName);
+
+                message.UserProperties.Add(Constants.InitializationVectorPropertyName, Convert.ToBase64String(iV));
+                message.UserProperties.Add(Constants.KeyNamePropertyName, secretName);
+
+                message.Body = await KeyVaultPlugin.Encrypt(Encoding.UTF8.GetString(message.Body), secret, iV);
+                return message;
+            }
+            catch (Exception ex)
+            {
+                throw new KeyVaultPluginException(Resources.BeforeMessageSendException, ex);
+            }
+        }
+
+        /// <summary>
+        /// The action performed after receiving a message from Service Bus. This method will load the KeyVault key and decrypt messages.
+        /// </summary>
+        /// <param name="message">The <see cref="Message"/> to be decrypted.</param>
+        /// <returns>The decrypted <see cref="Message"/>.</returns>
+        public override async Task<Message> AfterMessageReceive(Message message)
+        {
+            try
+            {
+                if (!message.UserProperties.ContainsKey(Constants.InitializationVectorPropertyName) || !message.UserProperties.ContainsKey(Constants.KeyNamePropertyName))
+                {
+                    return message;
+                }
+
+                var iVString = message.UserProperties[Constants.InitializationVectorPropertyName] as string;
+                var iV = Convert.FromBase64String(iVString);
+                var secretName = message.UserProperties[Constants.KeyNamePropertyName] as string;
+
+                var secret = await secretManager.GetHashedSecret(secretName);
+
+                var decryptedMessage = await KeyVaultPlugin.Decrypt(message.Body, secret, iV);
+
+                message.Body = Encoding.UTF8.GetBytes(decryptedMessage);
+                return message;
+            }
+            catch (Exception ex)
+            {
+                throw new KeyVaultPluginException(Resources.AfterMessageReceiveException, ex);
+            }
+        }
+
+        internal static Task<byte[]> GenerateInitializationVector()
+        {
+            byte[] initializationVector = null;
+            using (var aes = Aes.Create())
+            {
+                aes.GenerateIV();
+                initializationVector = aes.IV;
+            }
+            return Task.FromResult(initializationVector);
+        }
+
+        // Taken from the examples here: https://msdn.microsoft.com/en-us/library/system.security.cryptography.aes
+        internal static Task<byte[]> Encrypt(string payload, byte[] key, byte[] initializationVector)
```

Comment:
This should be refactored to just take a byte[], in case the user doesnt encode with UTF8